### PR TITLE
feat: app scaffolding manifest

### PR DESCRIPTION
**Pull request by jgabrielmaia**
- Support to initialize apps: creates cmfapp.json
- Support to pack app projects: generates app_manifest and icon
- Unit tests for init and pack operations
- Unit test for cmf app loading
- Unit test supporting icon utilities

### CODE REVIEW

File: cmf-cli/Constants/CliConstants.cs [hunk 1]
```diff
@@ -33,11 +33,36 @@ public static class CliConstants
         /// </summary>
         public const string DeploymentFrameworkManifestFileName = "manifest.xml";
 
+        /// <summary>
+        /// The deployment framework manifest template
```
Comment (pedrotinoco) [inline cmf-cli/Constants/CliConstants.cs:37]: change deployment to `app`, please

File: cmf-cli/resources/templateFiles/app_manifest.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<App packageId="$(id)"
```
Comment (pedrotinoco) [inline cmf-cli/resources/templateFiles/app_manifest.xml:2]: This must be `id`

File: core/Constants/CoreConstants.cs [hunk 3]
```diff
@@ -31,6 +31,11 @@ public static class CoreConstants
         /// </summary>
         public const string PackageJson = "package.json";
 
+        /// <summary>
+        /// The CMF package file name
```
Comment (pedrotinoco) [inline core/Constants/CoreConstants.cs:35]: Maybe change `package` to manifest?

File: core/Objects/CmfApp.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public static void SaveIcon(CmfApp cmfApp, string iconDestination)
+    {
+        string iconSource = cmfApp.Content.App.Image.File;
+
+        byte[] iconBytes = cmfApp.FileSystem.File.ReadAllBytes(iconSource);
+
+        cmfApp.FileSystem.File.WriteAllBytes(iconDestination, iconBytes);
+    }
```
Comment (pedrotinoco) [inline core/Objects/CmfApp.cs:141]: I'd say that both Save and SaveIcon need documentation

File: tests/Specs/AppIconUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        public void ValidPngIcon_ReturnsTrue()
+        {
+            var path = Path.Combine(TestFolder, PNG_ICON_PATH);
+
+            try
+            {
+                ImageCreator.CreatePngImage(Path.Combine(TestFolder, PNG_ICON_PATH), 100, 100);
+                
+                bool result = AppIconUtilities.IsIconValid(path);
+                Assert.True(result);
```
Comment (pedrotinoco) [inline tests/Specs/AppIconUtilities.cs:44]: All asserts should have error messages just so test tracking gets easier

Comment (pedrotinoco) [reply tests/Specs/AppIconUtilities.cs:44]: Please, check other cases like this one

File: tests/Specs/Init.cs [hunk 6]
```diff
@@ -22,7 +26,7 @@ public class Init
     {
         public Init()
         {
-                ExecutionContext.ServiceProvider = (new ServiceCollection())
+                Cmf.CLI.Core.Objects.ExecutionContext.ServiceProvider = (new ServiceCollection())
```
Comment (pedrotinoco) [inline tests/Specs/Init.cs:29]: Is this change really needed?

File: cmf-cli/Commands/init/InitCommand.cs [hunk 7]
```diff
@@ -174,7 +182,44 @@ public override void Configure(Command cmd)
...
+            cmd.AddOption(new Option<string>(
+                aliases: new[] { "--appLicensedApplication" },
+                description: "Use only if repository type is App. Licensed application"
+            ) { IsRequired = false });
+
+            cmd.AddOption(new Option<string>(
+                aliases: new[] { "--appIcon" },
+                description: "Use only if repository type is App. Application icon"
+            ) { IsRequired = false });
+
```
[the remark this answers is not in the document]
Comment (jgabrielmaia) [reply cmf-cli/Commands/init/InitCommand.cs:222]: Since it is an init file it should always be the default start, right?
Either 0.1.0 or 1.0.0

I can instead allow a version parameter with a default value

[the remark this answers is not in the document]
Comment (pedrotinoco) [reply cmf-cli/Commands/init/InitCommand.cs:222]: Yes, I understand your perspective and I subscribe it. But shouldn't we allow to define the AppVersion? Maybe @m-s- can help here

File: core/Objects/CmfApp.cs [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+{
+    [XmlAttribute("file")]
+    public string File { get; set; }
+}
+
+[XmlRoot("App")]
+public record CmfAppV1
+{
+    [XmlAttribute("id")] public string Id { get; set; }
+    [XmlAttribute("version")] public string Version { get; set; }
```
[the remark this answers is not in the document]
Comment (jgabrielmaia) [reply core/Objects/CmfApp.cs:49]: Even though we discussed using the Version object it seemed to me that it would bring complexity and inconsistency, since other classes/methods are still using string Version, such as Bump in PackageTypeHandler and CmfPackage

File: cmf-cli/Handlers/PackageType/PackageTypeHandler.cs [hunk 9]
```diff
@@ -264,6 +272,41 @@ internal virtual void GenerateDeploymentFrameworkManifest(IDirectoryInfo package
...
+        internal virtual void GenerateDeploymentFrameworkAppFiles(IDirectoryInfo packageOutputDir)
+        {
+            Log.Debug("Generating DeploymentFramework App manifest");
+            string path = Path.Combine(packageOutputDir.FullName, CliConstants.DeploymentFrameworkAppManifestFileName);
+
+            IFileInfo cmfAppFile = fileSystem.FileInfo.New(CliConstants.CmfAppFileName);
+
+            CmfApp cmfApp = CmfApp.Load(cmfAppFile, fileSystem);
+
+            if (cmfApp.Content.App.Image.File == string.Empty)
```
[the remark this answers is not in the document]
Comment (jgabrielmaia) [reply cmf-cli/Handlers/PackageType/PackageTypeHandler.cs:289]: Changed to string.IsNullOrWhiteSpace

File: cmf-cli/Handlers/PackageType/PackageTypeHandler.cs [hunk 10]
```diff
@@ -587,6 +630,11 @@ public virtual void Pack(IDirectoryInfo packageOutputDir, IDirectoryInfo outputD
 
             GenerateDeploymentFrameworkManifest(packageOutputDir);
 
+            if (ExecutionContext.Instance.ProjectConfig?.RepositoryType == RepositoryType.App)
```
[the remark this answers is not in the document]
Comment (jgabrielmaia) [reply cmf-cli/Handlers/PackageType/PackageTypeHandler.cs:633]: In what case the cmfapp.json would exist and not be an app?

File: core/Constants/CoreConstants.cs (hunk 3 above)
Comment (jgabrielmaia) [reply core/Constants/CoreConstants.cs:35]: In this case it is not the manifest yet, but cmfApp.json. Will keep it documented as CMF app file name

File: cmf-cli/Handlers/PackageType/PackageTypeHandler.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pedrotinoco) [reply cmf-cli/Handlers/PackageType/PackageTypeHandler.cs:633]: IMO, never. My question is: Why not just checking if the cmfapp.json exists, instead of checking the repo type?

[the remark this answers is not in the document]
Comment (jgabrielmaia) [reply cmf-cli/Handlers/PackageType/PackageTypeHandler.cs:633]: I chose this approach because it seemed cheaper than using the file system. But @m-s- can enlighten us in this matter

File: tests/Specs/Pack.cs [hunk 11]
```diff
@@ -316,6 +316,46 @@ public void Pack_SecurityPortal()
...
+
+            DirectoryInfo curDir = new(System.IO.Directory.GetCurrentDirectory());
+
+            Assert.True(Directory.Exists($"{dir}/Package"), "Package folder is missing");
+            Assert.True(File.Exists($"{dir}/Package/{packageName}"), "Zip package is missing");
+
+            List<string> entries = TestUtilities.GetFileEntriesFromZip($"{dir}/Package/{packageName}");
+            Assert.True(entries.HasAny(), "Zip package is empty");
+            Assert.True(entries.HasAny(entry => entry == "manifest.xml"), "Manifest file does not exist");
+            Assert.True(entries.HasAny(entry => entry == "app_manifest.xml"), "App manifest file does not exist");
```
Comment (m-s-) [inline tests/Specs/Pack.cs:355]: should be separated? seems like this is consumed by a different target system than the normal package manifest

**Review by m-s- — CHANGES_REQUESTED** [review]

**Review by m-s- — DISMISSED** [review]

**Comment by m-s-** [comment]
LGTM, will merge after 4.2.0 stable is out

File: core/Commands/TemplateCommand.cs [hunk 12]
```diff
@@ -146,6 +147,11 @@ protected async void ExecuteTemplate(string templateName, IReadOnlyCollection<st
                     throw new CliException(
                         $"Template generation failed with status {result.Status}: {result.ErrorMessage}");
                 }
+
+                if(ExecutionContext.Instance.ProjectConfig.RepositoryType == RepositoryType.App)
+                {
+                    Log.Debug($"Make sure you create a business project for your application");
```
Comment (pedrotinoco) [inline core/Commands/TemplateCommand.cs:153]: I'd change this message to something like: Apps need to have an ApplicationVersion project. Make sure you create at least one business project for your application using the addApplicationVersionAssembly

Comment (m-s-) [reply core/Commands/TemplateCommand.cs:153]: Also, verbosity should be at least Information, Debug isn't printed unless you specifically request it with --loglevel

**Review by m-s- — DISMISSED** [review]

**Review by m-s- — APPROVED** [review]
